models/Image.js

```diff
@@ -1,76 +1,155 @@
+}
+
+/**
+ *  Build resize metadata.
+ *
+ *  @param {string} filename - uploaded file name
+ *  @param {number} originWidth - image width
+ *  @param {number} originHeight - image height
+ *  @returns {Object}
+ */
+function getResizeMetadata(filename, originWidth, originHeight) {
+  const [filenameWithoutExt, ext] = filename.split('.');
+  const targets = {};
+
+  resizedOpts.forEach(resizedOpt => {
+    let [resolution, resizeWidth] = resizedOpt;
+    let resizeHeight = 0;
+
+    if (resizeWidth < originWidth) {
+      resizeHeight = Math.round((resizeWidth / originWidth) * originHeight);
+    } else {
+      resizeWidth = originWidth;
+      resizeHeight = originHeight;
+    }
+
+    targets[resolution] = {
+      url: `${resizeImageURLPrefix}${filenameWithoutExt}-${resolution}.${ext}`,
+      width: resizeWidth,
+      height: resizeHeight,
+    };
+  });
+
+  return targets;
+}
 
 Image.add({
-    image: {
-        type: Types.GcsImage,
-        initial: true,
-        autoCleanup: true,
-        datePrefix: 'YYYYMMDDHHmmss',
-        // TODO move these settings to config
-        bucket: bucket,
-        destination: 'images/',
-        publicRead: true,
-        resize: resizeImage,
-        resizeOpts: [{
-            target: 'desktop',
-            width: 2000,
-            height: null,
-            options: {}
-        }, {
-            target: 'tablet',
-            width: 1200,
-            height: null,
-            options: {}
-        }, {
-            target: 'mobile',
-            width: 800,
-            height: null,
-            options: {}
-        }, {
-            target: 'w400',
-            width: 400,
-            height: null,
-            options: {}
-        }, {
-            target: 'tiny',
-            width: 150,
-            height: null,
-            options: {}
-        }],
-        extractIPTC: extractIPTC
-    },
-    description: {
-        type: String,
-        index: true
-    },
-    copyright: {
-        type: Types.Select,
-        options: 'Creative-Commons, Copyrighted',
-        default: 'Copyrighted', index: true
-    },
-    keywords: {
-        type: String,
-        index: true
-    },
-    sale: {
-        type: Boolean,
-        index: true
+  image: {
+    type: Types.GcsImage,
+    initial: true,
+    autoCleanup: true,
+    datePrefix: 'YYYYMMDDHHmmss',
+    bucket: bucket,
+    destination: '/',
+    publicRead: false,
+    hooks: {
+      postUpload: (imageData) => {
+        if (typeof imageData !== 'object' || imageData === null) {
+          return Promise.resolve(imageData);
+        }
+
+        const { filepath, filename, height, width } = imageData;
+
+        if (typeof filename !== 'string' || filename === '') {
+          return Promise.resolve(imageData);
+        }
+
+        return publishResizeTasks(filename)
+          .catch(err => {
+            console.error(`Received error while triggering ResizeImage GC.`);
+            return Promise.reject(err);
```

Comment:
Do we want to chain the error here? Will it better to use `Promise.all` to run both `publishResizeTasks` & `extractIPTC` and annotate the error within the functions?

Reply:
If we use `Promise.all`, such as following demo codes,

```
Promise.all([
  publishResizeTasks(filename),
  extractIPTC(filepath),
]).catch(err => {
  // not easy to know which error occurs.
})
```
it's not easy for us to know which error occurs,
is it publishing error, 
or is it extracting iptc error?

Therefore, current way could provide more error information.

Reply:
> If we use `Promise.all`, such as following demo codes,
> 
> ```
> Promise.all([
>   publishResizeTasks(filename),
>   extractIPTC(filepath),
> ]).catch(err => {
>   // not easy to know which error occurs.
> })
> ```

That is why I suggest moving the error wrapping(i.e. `console.error("Received error while triggering ResizeImage GC.");` within each of the function. 

> 
> it's not easy for us to know which error occurs,
> is it publishing error,
> or is it extracting iptc error?
> 
> Therefore, current way could provide more error information.
IMO, the error should be only handled once including `console.log`. Multiple error information with the same root cause would probably result in bias and hard to uncover at the very first sight.

Reply:
Updated.